source/solvers/heat_transfer_assemblers.cc

```diff
@@ -523,3 +522,118 @@ HeatTransferAssemblerViscousDissipation<dim>::assemble_rhs(
 
 template class HeatTransferAssemblerViscousDissipation<2>;
 template class HeatTransferAssemblerViscousDissipation<3>;
+
+
+
+template <int dim>
+void
+HeatTransferAssemblerLaser<dim>::assemble_matrix(
+  HeatTransferScratchData<dim> & /*scratch_data*/,
+  StabilizedMethodsCopyData & /*copy_data*/)
+{}
+
+template <int dim>
+void
+HeatTransferAssemblerLaser<dim>::assemble_rhs(
+  HeatTransferScratchData<dim> &scratch_data,
+  StabilizedMethodsCopyData &   copy_data)
+{
+  // Laser parameters
+  const double concentration_factor = laser_parameters->concentration_factor;
+  const double laser_power          = laser_parameters->laser_power;
+  const double absorptivity         = laser_parameters->laser_absorptivity;
+  const double penetration_depth    = laser_parameters->penentration_depth;
+  const double laser_start_time     = laser_parameters->start_time;
+  const double laser_end_time       = laser_parameters->end_time;
+  const double beam_radius          = laser_parameters->beam_radius;
+  const double current_time = this->simulation_control->get_current_time();
+
+  if (current_time >= laser_start_time && current_time <= laser_end_time)
+    {
+      // Get laser path
+      Function<dim> &laser_path = *(laser_parameters->laser_path);
+      laser_path.set_time(current_time);
+
+      // Get laser lcoation
+      Point<dim> laser_location;
+      laser_location[0] =
+        laser_path.value(laser_location,
+                         laser_parameters->perpendicular_plane_coordinate_one);
+      laser_location[1] =
+        laser_path.value(laser_location,
+                         laser_parameters->perpendicular_plane_coordinate_two);
+
+      double laser_location_in_depth = 0.0;
+      if constexpr (dim == 3)
+        {
+          laser_location[2] =
+            laser_path.value(laser_location,
+                             laser_parameters->beam_direction_coordinate);
+          laser_location_in_depth = laser_location[2];
+        }
+
+      // Get laser location on the operation surface
+      Point<2> laser_location_on_surface;
```

Comment:
why is this always a point<2>. Should it not be a Point<dim-1> ???

Reply:
> For the calculation of the laser heat source term, we need the radial distance between the laser focal point (in a 2d plane perpendicular to the laser beam direction) and the quadrature points, and axial distance (in the emission direction) between the laser focal point and the quadrature points, separately. Hence, we store the laser position on the surface as a Point<2> to calculate of the radial distance, and laser position in the emission direction (laser_location_in_depth as a double) to calculate the axial distance.

Yeah but this does not work in a 2D simulation. In this case the point will be a 1D point and not a 2D point.
You need to make sure it's  consistent...

Reply:
It is revised according to our discussion :)